spring-cloud-gcp-pubsub-stream-binder/src/test/java/org/springframework/cloud/gcp/stream/binder/pubsub/PubSubEmulator.java

```diff
@@ -0,0 +1,160 @@
+/*
+ *  Copyright 2018 original author or authors.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.stream.binder.pubsub;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.junit.rules.ExternalResource;
+
+/**
+ * Rule for instantiating and tearing down a Pub/Sub emulator instance.
+ */
+public class PubSubEmulator extends ExternalResource {
+	private static final Path emulatorConfigPath = Paths.get(System.getProperty("user.home")).resolve(
+			Paths.get(".config", "gcloud", "emulators", "pubsub", "env.yaml"));
+
+	private static final Log LOGGER = LogFactory.getLog(PubSubEmulator.class);
+
+
+	// Binder to Pub/Sub emulator for use in individual tests.
+	private PubSubTestBinder binder;
+
+	// Reference to emulator instance, for cleanup.
+	private Process emulatorProcess;
+
+	// Hostname for cleanup, should always be localhost.
+	private String emulatorHost;
+
+	// Port for cleanup.
+	private String emulatorPort;
+
+	/**
+	 * Launches an instance of pubsub simulator, creates a binder.
+	 *
+	 * @throws Throwable for any failed setup.
+	 */
+	@Override
+	protected void before() throws Throwable {
+
+		this.emulatorProcess = new ProcessBuilder("gcloud", "beta", "emulators", "pubsub", "start").start();
```

Comment:
Does it matter that an emulator might already be running or multiple are allowed to run at the same time?

Reply:
Multiples are allowed, but in local testing, after a couple of localhost:NNNN instances, it will go into IPv6 mode (localhost:::1), which the binder can't handle.

So I'd rather take them down once the test completes.